File: core/shared/src/main/scala/diplomacy/rule/OrderResult.scala
package diplomacy.rule

import scala.scalajs.js.annotation.{ JSExportDescendentObjects, JSExport, JSExportAll }

import diplomacy.board.{ Power, MilitaryBranch }

@JSExportDescendentObjects @JSExportAll
sealed trait OrderResult[Power_ <: Power, MilitaryBranch_ <: MilitaryBranch, Order_ <: Order[Power_, MilitaryBranch_], Result_] {
  val target: Order_
  val result: Result_
}

@JSExport
object OrderResult {
  @JSExportAll
  final case class Executed[Power_ <: Power, MilitaryBranch_ <: MilitaryBranch, Order_ <: Order[Power_, MilitaryBranch_], Result_](
    target: Order_, result: Result_
  ) extends OrderResult[Power_, MilitaryBranch_, Order_, Result_]

  @JSExportAll
  final case class Replaced[Power_ <: Power, MilitaryBranch_ <: MilitaryBranch, Order_ <: Order[Power_, MilitaryBranch_], Result_](
    target: Order_, invalidReason: InvalidOrderMessage, replacedBy: Order_, result: Result_
  ) extends OrderResult[Power_, MilitaryBranch_, Order_, Result_]
}

File: src/rule/order-result.ts
import { Order } from "./order"

export interface OrderResult<Power, MilitaryBranch, Result> {
  /**
   * The target order.
   */
  target: Order<Power, MilitaryBranch>
  /**
   * The result of the target.
   */
  result: Result
}

/**
 * Result of an order execution.
 */
export class Executed<Power, MilitaryBranch, Result>
  implements OrderResult<Power, MilitaryBranch, Result> {
  /**
   * @param target The target order.
   * @param result The result of the target.
   */
  constructor (public target: Order<Power, MilitaryBranch>, public result: Result) {}
}
/**
 * Result of an order execution. It is used when the original order is replaced.
 */
export class Replaced<Power, MilitaryBranch, Result>
  implements OrderResult<Power, MilitaryBranch, Result> {
  /**
   * @param target The target order.
   * @param result The result of the target.
   * @param invalidReason The reason why the target is replaced
   * @param replacedBy The order that replaces the target.
   */
  constructor (
    public target: Order<Power, MilitaryBranch>,
    public invalidReason: string,
    public replacedBy: Order<Power, MilitaryBranch>,
    public result: Result
  ) {}
}
